extensions/s3/src/main/java/io/deephaven/extensions/s3/S3SeekableByteChannel.java

```diff
@@ -0,0 +1,324 @@
+            final int readAheadFragmentIndex = i + currFragmentIndex + 1;
+            final CompletableFuture<ByteBuffer> readAheadFragmentFuture =
+                    context.getCachedFuture(readAheadFragmentIndex);
+            if (readAheadFragmentFuture == null) {
+                context.setFragmentContext(readAheadFragmentIndex, computeFragmentFuture(readAheadFragmentIndex));
+            }
+        }
+
+        // Wait till the current fragment is fetched
+        final ByteBuffer currentFragment;
+        try {
+            currentFragment = currFragmentFuture.get(readTimeout.toNanos(), TimeUnit.NANOSECONDS);
+        } catch (final InterruptedException | ExecutionException | TimeoutException | CancellationException e) {
+            final String operation =
+                    "fetching fragment " + currFragmentIndex + " for file " + key + " in S3 bucket " + bucket;
+            throw handleS3Exception(e, operation);
+        }
+
+        // Copy the bytes from fragment from the offset up to the min of remaining fragment and destination bytes.
+        // Therefore, the number of bytes read by this method can be less than the number of bytes remaining in the
+        // destination buffer.
+        final int fragmentOffset = (int) (localPosition - (currFragmentIndex * fragmentSize));
+        currentFragment.position(fragmentOffset);
+        final int limit = Math.min(currentFragment.remaining(), destination.remaining());
+        final int originalBufferLimit = currentFragment.limit();
+        currentFragment.limit(currentFragment.position() + limit);
+        destination.put(currentFragment);
+        // Need to reset buffer limit, so we can read from the same buffer again in future
+        currentFragment.limit(originalBufferLimit);
+        position = localPosition + limit;
+        return limit;
+    }
+
+    private int fragmentIndexForByteNumber(final long byteNumber) {
+        return Math.toIntExact(byteNumber / fragmentSize);
+    }
+
+    private CompletableFuture<ByteBuffer> computeFragmentFuture(final int fragmentIndex) {
+        final long readFrom = (long) fragmentIndex * fragmentSize;
+        final long readTo = Math.min(readFrom + fragmentSize, size) - 1;
+        final String range = "bytes=" + readFrom + "-" + readTo;
+        return s3AsyncClient.getObject(builder -> builder.bucket(bucket).key(key).range(range),
+                new ByteBufferAsyncResponseTransformer<>(fragmentSize));
+    }
+
+    private UncheckedDeephavenException handleS3Exception(final Exception e, final String operationDescription) {
+        if (e instanceof InterruptedException) {
+            Thread.currentThread().interrupt();
+            return new UncheckedDeephavenException("Thread interrupted while " + operationDescription, e);
+        } else if (e instanceof ExecutionException) {
+            return new UncheckedDeephavenException("Execution exception occurred while " + operationDescription, e);
```

Comment:
I think we might want to wrap ExecutionException as IOException since read interface can support that.

Reply:
I can do that but that will kind of make this part of the code ugly because right now we are returning `UncheckedDeephavenException` in all cases. After the change, `ExecutionException` will have to be handled outside along with the calling code at two places. Do you think I should go ahead?

Reply:
Should we use `IOException` for all of these?

Reply:
I made this change, let me know if it looks better.